### DISCUSSION TITLE
Use kkt operator to get Stationarity and Complementarity constraints.

### DISCUSSION THREAD

**Discussion by asxadf** [yalmip/YALMIP#1056] [Optimization modelling] [answered]
Hi Dr. Löfberg,

Thank you for hosting such a wonderful community for us. And I appreciate your brilliant creation YALMIP.

I want to use the KKT conditions to represent a lower level of bilevel programming, and further use Big-M reformulation to handle the bilinear term.

I thought YALMIP's kkt operator can do it. After checking the Output KKTsystem and Output Details, I am confused about some points:

1. For the Stationarity of KKT conditions, I didn't find the matrix in the Details. How can I get it? I thought I could get the matrix of its coefficients.

2. For the Complementarity Constraints of KKT conditions, I noticed the kkt operator has considered it in the KKTsystem. But I thought kkt operator handle it in a bilinear way. Even though GUROBI can solve the problem with bilinear terms, but it is computationally intractable in a large-size problem. So I want to find some way to get the Complementarity Constraints, so that I can manually use Big-M to linearize it.

I would like to appreciate you if you could share any ideas with me.

Best regards,
Xianbang

I am using the following example to learn the kkt operator:

```
sdpvar x1 x2 y1 y2 y3;

Outer_OBJ = -8*x1 - 4*x2 + 4*y1 - 40*y2 - 4*y3;
Inner_OBJ =    x1 + 2*x2 +   y1 +    y2 + 2*y3;

Outer_CON = [x1 >= 0]...
          + [x2 >= 0];
      
      
Inner_CON = [y1 >= 0]...
          + [y2 >= 0]...
          + [y3 >= 0]...
          + [     -   y1 +   y2 +     y3 - 1 <= 0]...
          + [2*x1 -   y1 + 2*y2 - 0.5*y3 - 1 <= 0]...
          + [2*x2 + 2*y1 -   y2 - 0.5*y3 - 1 <= 0];

[KKT_CON,details] = kkt(Inner_CON,Inner_OBJ, [x1 x2]);
```

**Comment by johanlofberg** [comment, marked as answer]
Details contains the data defining the problem, not the KKT conditions for the problem. 

kkt only returns a complementarity constraint. How this is used for a specific solver depends on the capabilities of the solver. In the case of gurobi (or any other integer solver), it is converted to a big-M model. If you use fmincon, it is written as a bilinear constraint, and if you use knitro, the complementarity constraint is used directly

To extract the components of a complementarity constraint  `F = complements(x >= 0, A*x <= b)` you can get x and b-A*x via

```
Z = sdpvar(F);
s1 = Z(:,1);
s2 = Z(:,2);
```

**Reply by asxadf** [reply]
Thanks for your comments. I just made some progress.

For point 1, I think using element 2 of the KKTSystem is fine. Because the kkt Operator has derived the stationarity, which can be treated as constraints directly. Such as the following:
`KKT_Stationary = KKT_Stationary + KKT_CON(2);`

For point 2, if I want to do the Big-M manually, one option could be simply repeating inequalities and complementing it with the corresponding dual variable in KKTDetails. Such as the following:
`KKT_COMPLEMENT = KKT_COMPLEMENT + complements(     -   y1 +   y2 +     y3 - 1 <= 0 , details.dual(1) >= 0);`
For this option, I have to correctly and carefully identify the dual variable for each inequality.

I write the following code, it reports the same result compared to solvebilevel operator. Hopefully, I am using kkt operator correctly.
```
sdpvar x1 x2 y1 y2 y3;

Outer_OBJ = -8*x1 - 4*x2 + 4*y1 - 40*y2 - 4*y3;
Inner_OBJ =    x1 + 2*x2 +   y1 +    y2 + 2*y3;

Outer_CON = [x1 >= 0]...
          + [x2 >= 0];
      
Inner_CON = [];
Inner_CON = Inner_CON + [     -   y1 +   y2 +     y3 - 1 <= 0];
Inner_CON = Inner_CON + [2*x1 -   y1 + 2*y2 - 0.5*y3 - 1 <= 0];
Inner_CON = Inner_CON + [2*x2 + 2*y1 -   y2 - 0.5*y3 - 1 <= 0];
Inner_CON = Inner_CON + [-y1 <= 0];
Inner_CON = Inner_CON + [-y2 <= 0];
Inner_CON = Inner_CON + [-y3 <= 0];

[KKT_CON,details] = kkt(Inner_CON,Inner_OBJ, [x1 x2]);

%% --------------------------------- C1 -------------------------------- %%
KKT_Stationary = [];
KKT_Stationary = KKT_Stationary + KKT_CON(2);
%% --------------------------------- C2 -------------------------------- %%
KKT_Primal_FEA = [];
KKT_Primal_FEA = KKT_Primal_FEA + [details.A*details.primal <= details.b];
%% --------------------------------- C3 -------------------------------- %%
KKT_Dual_FEA = [];
KKT_Dual_FEA = KKT_Dual_FEA + [details.dual >= 0];
%% --------------------------------- C4 -------------------------------- %%
KKT_COMPLEMENT = [];
KKT_COMPLEMENT = KKT_COMPLEMENT + complements(     -   y1 +   y2 +     y3 - 1 <= 0 , details.dual(1) >= 0);
KKT_COMPLEMENT = KKT_COMPLEMENT + complements(2*x1 -   y1 + 2*y2 - 0.5*y3 - 1 <= 0 , details.dual(2) >= 0);
KKT_COMPLEMENT = KKT_COMPLEMENT + complements(2*x2 + 2*y1 -   y2 - 0.5*y3 - 1 <= 0 , details.dual(3) >= 0);
KKT_COMPLEMENT = KKT_COMPLEMENT + complements(     -   y1                     <= 0 , details.dual(4) >= 0);
KKT_COMPLEMENT = KKT_COMPLEMENT + complements(            -   y2              <= 0 , details.dual(5) >= 0);
KKT_COMPLEMENT = KKT_COMPLEMENT + complements(                   -     y3     <= 0 , details.dual(6) >= 0);
%% ------------------------------- Solve ------------------------------- %%
KKT_CON_Handy  = KKT_Stationary + KKT_Primal_FEA + KKT_Dual_FEA + KKT_COMPLEMENT; 
ops = sdpsettings('solver', 'gurobi');
sol = optimize([Outer_CON, KKT_CON_Handy], Outer_OBJ, ops);
```

**Reply by johanlofberg** [reply]
I don't understand what you are doing. Why are you adding KKT_COMPLEMENT manually when you already called kkt and thus have it available in KKT_CON(1). YALMIP will use the same big-M modelling to represent the complementarity in both models

**Reply by johanlofberg** [reply]
...and the primal constraints and dual non-negativity is in the complementarity definition, so your whole code is equivalent to 

```
KKT_CON_Handy  = KKT_CON; 
ops = sdpsettings('solver', 'gurobi');
sol = optimize([Outer_CON, KKT_CON_Handy], Outer_OBJ, ops);
```

**Reply by asxadf** [reply]
Hi Dr. Löfberg,

Thanks for your comment.

Yes, the code is equivalent to what you attached, in which KKT operator directly gives me the whole KKT system. And this is very convenient in a small case.

However, in a large case, I want to get full control of the KKT conditions, especially the control of complementary constraints, so that I can improve the computation performance.

According to your view [https://yalmip.github.io/tutorial/bigmandconvexhulls](url), the computation performance is related to the M value, thus I want to select a specific M for different constraints. One option is using KKT operator to get the raw KKT system and then manually handling the complementary constraints.

The following code may show my idea:
```
sdpvar x1 x2 y1 y2 y3 l1 l2 l3 l4 l5 l6;
binvar u1 u2 u3 u4 u5 u6;

M1 = 1000;
M2 = 1000;
M3 = 1000;
M4 = 1000;
M5 = 1000;
M6 = 1000;

m1 = 1000;
m2 = 1000;
m3 = 1000;
m4 = 1000;
m5 = 1000;
m6 = 1000;

OBJ = -8*x1 - 4*x2 + 4*y1 - 40*y2 - 4*y3;
CON = [];
% Outer Constraints
CON = CON + [x1 >= 0];
CON = CON + [x2 >= 0];
% Stationarity
CON = CON + [1 - l1 -     l2 +   2*l3 - l4 == 0];
CON = CON + [1 + l1 +   2*l2 -     l3 - l5 == 0];
CON = CON + [2 + l1 - 0.5*l2 - 0.5*l3 - l6 == 0];
% Primal Feasibility
CON = CON + [y1 >= 0];
CON = CON + [y2 >= 0];
CON = CON + [y3 >= 0];
CON = CON + [         -   y1 +   y2 +     y3 - 1  <= 0];
CON = CON + [    2*x1 -   y1 + 2*y2 - 0.5*y3 - 1  <= 0];
CON = CON + [    2*x2 + 2*y1 -   y2 - 0.5*y3 - 1  <= 0];
% Dual Feasibility
CON = CON + [l1 >= 0];
CON = CON + [l2 >= 0];
CON = CON + [l3 >= 0];
CON = CON + [l4 >= 0];
CON = CON + [l5 >= 0];
CON = CON + [l6 >= 0];
% Complement    
CON = CON + [l1 <= u1*M1];
CON = CON + [l2 <= u2*M2];
CON = CON + [l3 <= u3*M3];
CON = CON + [l4 <= u4*M4];
CON = CON + [l5 <= u5*M5];
CON = CON + [l6 <= u6*M6];
CON = CON + [     -   y1 +   y2 +     y3 - 1 >= (u1 - 1)*m1];
CON = CON + [2*x1 -   y1 + 2*y2 - 0.5*y3 - 1 >= (u2 - 1)*m2];
CON = CON + [2*x2 + 2*y1 -   y2 - 0.5*y3 - 1 >= (u3 - 1)*m3];
CON = CON + [     -   y1                     >= (u4 - 1)*m4];
CON = CON + [            -   y2              >= (u5 - 1)*m5];
CON = CON + [                   -     y3     >= (u6 - 1)*m6];

ops = sdpsettings('solver', 'gurobi');
sol = optimize(CON, OBJ, ops);
```

In the application, I want to use KKT operator to export the `Stationarity`, `Primal Feasibility`, and `Dual Feasibility`, instead of encoding them manually above. For the `Complement`, I want to manually encode it so that I can fully control the Big-M reformulation, instead of using built-in logic programming or the `complements` operator. Hopefully, the full control could improve the computation performance.

Best regards,
Xianbang

**Reply by johanlofberg** [reply]
note that kkt derives bounds on primals and duals 8when possible), thus allowing you to derive better values than your magic constant 1000

**Reply by asxadf** [reply]
Yes, I noticed that KKT operator can provide me with the bounds in this case.

But in my case (14-bus unit commitment), the derivation is very slow (I am not sure about the reasons), so I turned it off.

As your comments at [https://yalmip.github.io/tutorial/bigmandconvexhulls](url) probably I should get insights into my specific UC problem and find suitable M values.

**Reply by asxadf** [reply]
> 

By the way, I would like to confirm the term "bounds on primals and duals". Is this "bounds" essentially the M value?

**Reply by johanlofberg** [reply]
No, but to derive bounds on b-Ax for the big-M modelling, you first need bounds on x. For the dual, the bound on the dual would effectively be the big-M value

**Reply by asxadf** [reply]
Got it. Thank you for your reply!

**Reply by asxadf** [reply]
Hi Dr. Löfberg,

I have tried the manual KKT in my UC case today. The most complicated part is the `Complement`: I have to carefully identify the dual variables in `details.dual` for each inequality. This is error-prone and very cumbersome. So I would like to know whether KKT operator support assigning a specific M for each dual variable. If so, I can avoid the manual process but still fully control the Big-M bound.

**Reply by johanlofberg** [reply]
Deriving big-M constants for the implications for duals to be zero simply means you have to derive upper bounds on the individual duals (which is hard here, the kkt operator fails completely despite running a full LP-based bound propagation, the general problem of deriving bounds on duals can be shown to be as hard as solving the bilevel problem to begin with). You can of course use (you should!) use different big-M constants on every single logic model, and if you ask YALMIP to derive these big-M models it will use all bounds you have added on the involved variables.

**Reply by johanlofberg** [reply]
and what is complicated about the big-M? The standard model is simply

```
d = binvar(length(details.dual),1);
implies(d, details.dual == 0) + implies(1-d, A*details.primals <= details.b)
```

which manually would be (with carefully derived and optimized big-M constant)

```
[s <= M.*d, A*details.primals - details.b <= M.*(1-d)]
```

**Reply by asxadf** [reply]
Thanks for your reply.

I am impressed with the deriving big-M function of the KKT operator. According to its function description, it could give me a small possible but large enough M, is that right? However, in my case (a 14-bus unit commitment), the deriving is too slow. Thus I am not sure whether it is worth giving the KKT operator enough time to do the deriving. As I mentioned, if it could give a good M, I think it is worth it.

Regarding the manual `Complement`, I just realized that the `A` and `b` in `details` offer a gateway to do it. Such a smart operator!

**Reply by asxadf** [reply]
Even though `implies`, `complements`, and `|` can help me do the Big-M reformulation automatically, I am still concerned about one point: Not free lunch, so the compared to`[s <= M.*d, A*details.primals - details.b <= M.*(1-d)]` whether `implies(d, details.dual == 0) + implies(1-d, A*details.primals <= details.b)` will introduce more constraints thus increasing the model size?

Also, if I use `implies`, `complements`, and `|`, seems like I cannot select specific M for them.
